lib/sshutils/sftp/sftp.go

```diff
@@ -0,0 +1,472 @@
+	// Create creates a new file
+	Create(ctx context.Context, path string, mode os.FileMode) (io.WriteCloser, error)
+	// MkDir creates a directory
+	// sftp.Client.Mkdir does not take an os.FileMode, so this can't either
+	Mkdir(ctx context.Context, path string, mode os.FileMode) error
+	// Chmod sets file permissions
+	Chmod(ctx context.Context, path string, mode os.FileMode) error
+	// Chtimes sets file access and modification time
+	Chtimes(ctx context.Context, path string, atime, mtime time.Time) error
+}
+
+// CreateUploadConfig returns a Config ready to upload files
+func CreateUploadConfig(src []string, dst string, opts Options) (*Config, error) {
+	for _, srcPath := range src {
+		if srcPath == "" {
+			return nil, trace.BadParameter("source path is empty")
+		}
+	}
+	if dst == "" {
+		return nil, trace.BadParameter("destination path is empty")
+	}
+
+	c := &Config{
+		srcPaths: src,
+		dstPath:  dst,
+		srcFS:    &localFS{},
+		dstFS:    &remoteFS{},
+		opts:     opts,
+	}
+	c.setDefaults()
+
+	return c, nil
+}
+
+// CreateDownloadConfig returns a Config ready to download files
+func CreateDownloadConfig(src, dst string, opts Options) (*Config, error) {
+	if src == "" {
+		return nil, trace.BadParameter("source path is empty")
+	}
+	if dst == "" {
+		return nil, trace.BadParameter("destination path is empty")
+	}
+
+	c := &Config{
+		srcPaths: []string{src},
+		dstPath:  dst,
+		srcFS:    &remoteFS{},
+		dstFS:    &localFS{},
+		opts:     opts,
+	}
+	c.setDefaults()
+
+	return c, nil
+}
+
+// setDefaults sets default values
+func (c *Config) setDefaults() {
+	logger := c.Log
+	if logger == nil {
+		logger = log.StandardLogger()
+	}
+	c.Log = logger.WithFields(log.Fields{
+		trace.Component: "SFTP",
+		trace.ComponentFields: log.Fields{
+			"SrcPaths":      c.srcPaths,
+			"DstPath":       c.dstPath,
+			"Recursive":     c.opts.Recursive,
+			"PreserveAttrs": c.opts.PreserveAttrs,
+		},
+	})
+}
+
+// TransferFiles transfers files from the configured source paths to the
+// configured destination path over SFTP
+func (c *Config) TransferFiles(ctx context.Context, sshClient *ssh.Client) error {
+	sftpClient, err := sftp.NewClient(sshClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := c.initFS(ctx, sshClient, sftpClient); err != nil {
+		return trace.Wrap(err)
+	}
+
+	transferErr := c.transfer(ctx)
+	closeErr := sftpClient.Close()
+	if transferErr != nil {
+		return trace.Wrap(transferErr)
+	}
+
+	return trace.Wrap(closeErr)
+}
+
+// initFS ensures the source and destination filesystems are ready to transfer
+func (c *Config) initFS(ctx context.Context, sshClient *ssh.Client, client *sftp.Client) error {
+	var haveRemoteFS bool
+
+	srcFS, srcOK := c.srcFS.(*remoteFS)
+	if srcOK {
+		srcFS.c = client
+		haveRemoteFS = true
+	}
+	dstFS, dstOK := c.dstFS.(*remoteFS)
+	if dstOK {
+		dstFS.c = client
+		haveRemoteFS = true
+	}
+
+	if haveRemoteFS {
+		if c.getHomeDir == nil {
+			c.getHomeDir = func() (_ string, err error) {
```

Comment:
Why do we need to cache it? What if the directory changes between calls `usermod -m /home/foo bar` ?

Reply:
We don't have to cache it, but I wanted to avoid extra teleport re-execs. Also, if the remote user's home dir is changed between calls, one or more paths will be using the old home dir and will likely cause the transfer to fail. Given that, do you still think I should not cache the home dir?

Reply:
Nevermind, my logic is flawed... currently only one remote path is ever allowed, so caching is useless. Good catch! Removed.